app/models/grda_warehouse/census/census_builder.rb

```diff
@@ -6,43 +6,106 @@
+
 # Builds and populates the nightly_census_by_projects table.
-# Processes service and inventory data in date range batches.
 # Entry point: GrdaWarehouse::Tasks::CensusImport#run!
 module GrdaWarehouse::Census
   class CensusBuilder
-    def create_census(start_date, end_date)
-      batch_start_date = start_date.to_date
-      while batch_start_date <= end_date.to_date
-        # Batches are 1 month, or to the end_date if closer
-        batch_end_date = [batch_start_date + 1.years, end_date].min
-
-        GrdaWarehouseBase.transaction do
-          batch_by_project = ProjectBatch.new(batch_start_date, batch_end_date)
-          batch_by_project.build_census_batch
-
-          # Remove any existing census data for the batch range
-          ByProject.where(date: batch_start_date..batch_end_date).delete_all
-
-          # Save the new batch
-          headers = nil
-          values = []
-          batch_by_project.by_count.values.flat_map do |project|
-            headers = project.values.first.attributes.except('id').keys if headers.blank?
-            project.values.each do |day|
-              row = day.attributes.except('id')
-              row['created_at'] = Time.now
-              row['updated_at'] = Time.now
-              values << row.values
-            end
-            # project.values.each(&:save)
-          end
-          ByProject.new.insert_batch(ByProject, headers, values, transaction: false, batch_size: 500)
+    def self.call(...)
+      new(...).call
+    end
+
+    attr_accessor :start_date, :end_date, :progress
+    def initialize(start_date, end_date, progress: false)
+      self.start_date = start_date
+      self.end_date = end_date
+      self.progress = progress
+    end
+
+    def call
+      bar = new_progress_bar(relevant_projects.count) if progress
+      relevant_projects.find_each do |project|
+        rows = project_rows(project)
+        ByProject.transaction do
+          ByProject.where(project: project).delete_all
+          ByProject.import!(rows) if rows.any?
         end
+        # bar&.puts("Project ID #{project.id}, rows: #{rows.size}")
+        bar&.increment!(1)
+      end
+    end
+
+    protected
 
-        # Move batch forward
-        batch_start_date = batch_end_date + 1.day
+    def populations
+      @populations ||= GrdaWarehouse::Census.census_populations.map { |p| p[:population] }.uniq
+    end
+
+    def relevant_projects
+      GrdaWarehouse::Hud::Project
+    end
+
+    def project_rows(project)
+      merged = {}
+      populations.each do |population|
+        project_population_counts(project, population).each do |date, count|
+          merged[date] ||= { project_id: project.id }
+          merged[date][population] = count
+        end
       end
+      inventories = project.inventories.within_range(start_date..end_date)
+      merged.keys.each do |date|
+        inventories.each do |inventory|
+          next unless inventory_active_on_date?(inventory, date)
```

Comment:
I'm sure there's an optimization to be had to reduce the number of comparisons

Reply:
You could fetch the dates within the range where the inventory was active and convert it to a set, then it would just be a lookup of the date in question in the pre-calculated set.